udp_muxed_conn.go

```diff
@@ -28,52 +34,62 @@ type udpMuxedConn struct {
 	// Remote addresses that we have sent to on this conn
 	addresses []ipPort
 
-	// Channel holding incoming packets
-	buf        *packetio.Buffer
-	closedChan chan struct{}
-	closeOnce  sync.Once
-	mu         sync.Mutex
+	// FIFO queue holding incoming packets
+	bufHead, bufTail *bufferHolder
+	notify           chan struct{}
+	closedChan       chan struct{}
+	state            udpMuxedConnState
+	mu               sync.Mutex
 }
 
 func newUDPMuxedConn(params *udpMuxedConnParams) *udpMuxedConn {
-	p := &udpMuxedConn{
+	return &udpMuxedConn{
 		params:     params,
-		buf:        packetio.NewBuffer(),
+		notify:     make(chan struct{}, 1),
 		closedChan: make(chan struct{}),
 	}
-
-	return p
 }
 
 func (c *udpMuxedConn) ReadFrom(b []byte) (n int, rAddr net.Addr, err error) {
-	buf := c.params.AddrPool.Get().(*bufferHolder) //nolint:forcetypeassert
-	defer c.params.AddrPool.Put(buf)
-
-	// Read address
-	total, err := c.buf.Read(buf.buf)
-	if err != nil {
-		return 0, nil, err
-	}
-
-	dataLen := int(binary.LittleEndian.Uint16(buf.buf[:2]))
-	if dataLen > total || dataLen > len(b) {
-		return 0, nil, io.ErrShortBuffer
-	}
+	for {
+		c.mu.Lock()
+		for c.bufTail != nil {
```

Comment:
just a style comment, looks like this can be `if c.bufTail != nil`. I thought this loop around based on `for`, but looks like it does not.